screensaverd.c

```diff
@@ -0,0 +1,220 @@
+	vprintf(format, ap);
+	printf(": %s\n", strerror(errno));
+	va_end(ap);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static void
+close_fds(int fds[], int num)
+{
+	int i;
+
+	for (i = 0; i < num; i++)
+		close(fds[i]);
+}
+
+/* ------------------------------------------------------------------------ */
+
+/* Open all /dev/input/event* files. */
+static int
+open_fds(int fds[], int *num, int max_num)
+{
+	DIR *dir;
+	struct dirent *d;
+
+	if (!(dir = opendir(DEV_INPUT_DIR))) {
+		perrorf("Can't open directory %s", DEV_INPUT_DIR);
+		return -1;
+	}
+
+	while ((d = readdir(dir))) {
+		char name[256];
+
+		if (strncmp(d->d_name, EVENT_PREFIX, strlen(EVENT_PREFIX)))
+			continue; /* Not /dev/input/event* file */
+
+		snprintf(name, sizeof(name), "%s/%s", DEV_INPUT_DIR,
+			 d->d_name);
+		if ((fds[*num] = open(name, O_RDONLY)) == -1) {
+			perrorf("Can't open input device %s", name);
+			close_fds(fds, *num);
+			closedir(dir);
+			return -1;
+		}
+
+		if (++*num >= max_num)
+			break;
+	}
+
+	closedir(dir);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+sysfs_write_int(const char *fname, int val)
+{
+	FILE *f;
+
+	if (!(f = fopen(fname, "w"))) {
+		perrorf("Can't open \"%s\" for writing", fname);
+		return -1;
+	}
+
+	fprintf(f, "%d\n", val);
+	fclose(f);
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_on(void)
+{
+	return sysfs_write_int(DISPLAY_CONTROL, 0);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+turn_display_off(void)
+{
+	return sysfs_write_int(DISPLAY_CONTROL, 1);
+}
+
+/* ------------------------------------------------------------------------ */
+
+static int
+handle_events(int fd)
+{
+	char buf[BUF_SIZE];
+
+	/* Read and ignore event data if OK. */
+	if (read(fd, buf, sizeof(buf)) <= 0) {
+		perrorf("Error or EOF on read");
+		return -1;
+	}
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+void
+signal_handler(int sig UNUSED)
+{
+	running = 0;
+}
+
+/* ------------------------------------------------------------------------ */
+
+int
+main(void)
+{
+	int fds[MAX_DEVICES], num_fds = 0, ret = 0;
+
+	if (open_fds(fds, &num_fds, MAX_DEVICES) == -1)
+		return 1;
+
+	signal(SIGINT,  signal_handler);
+	signal(SIGTERM, signal_handler);
+
+	while (running) { /* Main loop */
+		int i, rv;
+		fd_set rfds;
+		struct timeval tv;
+
+		FD_ZERO(&rfds);
+		for (i = 0; i < num_fds; i++)
+			FD_SET(fds[i], &rfds);
+
+		/* Wait up to five seconds. */
+		tv.tv_sec  = DISPLAY_OFF_TIME;
+		tv.tv_usec = 0;
+
+		rv = select(num_fds, &rfds, NULL, NULL, &tv);
+		if (rv > 0) {
+			for (i = 0; i < num_fds; i++)
+				if (FD_ISSET(fds[i], &rfds) &&
+				    handle_events(fds[i]) == -1) {
+					running = 0;
+					break;
+				}
+
+			turn_display_on();
```

Comment:
Probably does not matter too much, but since the turn_display_on/off -helpers do not remember current state, we end up doing repeated sysfs writes especially from here.